processing/src/test/java/org/apache/druid/collections/CircularListTest.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.collections;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Set;
+
+public class CircularListTest
+{
+  @Test
+  public void testIterateInNaturalOrder()
+  {
+    final Set<String> input = ImmutableSet.of("b", "a", "c");
+    final CircularList<String> circularList = new CircularList<>(input, Comparator.naturalOrder());
+    final List<String> observedElements = new ArrayList<>();
+    int cnt = 0;
+    for (String x : circularList) {
+      observedElements.add(x);
+      if (++cnt >= input.size()) {
+        break;
+      }
+    }
+    Assert.assertEquals(ImmutableList.of("a", "b", "c"), observedElements);
+  }
+
+  @Test
+  public void testIterateInReverseOrder()
```

Comment:
Should we have a test where we have two `for` loops? We break the first in the middle of the cycle and ensure that the second resumes from where we started.
Or is that not a valid use case? 😛 

I think your current code will maintain the order with two or more `for` loops since the current position is maintained in the `CircularList` itself rather than inside the returned iterator.

Either way, it would be nice to call out the behaviour in the javadoc and have a test for it.

Reply:
Added a test and updated the javadoc to highlight the resume behavior.